sources/packages/api/src/services/student/student.service.ts

```diff
@@ -206,6 +225,97 @@ export class StudentService extends RecordDataModelService<Student> {
     return this.internalCreateStudent(userInfo, studentInfo, auditUserId);
   }
 
+  /**
+   * Search by an existing student using the SIN number and validates
+   * if the student has valid personal data that ensures the same identity.
+   * Case the SIN is present but it is not considered valid, exceptions will
+   * be thrown with the specific error types.
+   * @param studentSIN SIN to be searched.
+   * @param userInfo personal student data to be validated.
+   * @returns existing and validated student.
+   */
+  private async getExistingStudentForAccountCreation(
+    studentSIN: string,
+    userInfo: UserInfoMatchData,
+  ): Promise<Student> {
+    // Check if there are students with the same SIN.
+    const existingStudents = await this.getStudentsBySIN(studentSIN);
+    if (existingStudents.length > 1) {
+      const logMessage = "More than one student found for the provided SIN.";
+      this.logger.warn(logMessage);
+      throw new CustomNamedError(
+        logMessage,
+        STUDENT_ACCOUNT_CREATION_MULTIPLES_SIN_FOUND,
+      );
+    }
+    // Check if a student with the same SIN exists and has the same personal data.
+    if (existingStudents.length === 1) {
+      this.logger.warn("Found one student with the same SIN.");
+      const [studentFound] = existingStudents;
+      if (!this.studentMatchesUserInfo(studentFound, userInfo)) {
+        const logMessage =
+          "Student personal data does not match while trying to link the user to an existing student account.";
+        this.logger.warn(logMessage);
+        throw new CustomNamedError(
+          logMessage,
+          STUDENT_ACCOUNT_CREATION_FOUND_SIN_WITH_MISMATCH_DATA,
+        );
+      }
+      // Found the student with same SIN and personal data.
+      this.logger.warn(
+        "Student personal data matches. The student will be associated with the user.",
+      );
+      return studentFound;
+    }
+
+    return null;
+  }
+
+  /**
+   * Check if a student has a first name, last name and
+   * date of birth that matches with the provided data.
+   * @param student student to be verified.
+   * @param matchData data to be verified.
+   * @returns true if first name, last name and date of birth are the
+   * same, otherwise, false.
+   */
+  private studentMatchesUserInfo(
+    student: Student,
+    matchData: UserInfoMatchData,
+  ): boolean {
+    return (
+      dayjs(student.birthDate).isSame(matchData.birthdate) &&
+      // Using double equals (==) to consider null and undefined comparison as valid.
+      student.user.firstName?.toLowerCase() ==
```

Comment:
There might be need to trim here as well.

Reply:
The data being compared here came from the token and from our DB, the expectation is that the data would be sanitized already. If trim is needed it must be while saving the data, IMO.
For BCSC users we always use the data from the token (which must be sanitized), because we have manually inputted data from User/Ministry now for BCeID we can do it on the saving part. Does it make sense?

Reply:
Thanks for the explanation.

Reply:
Added trim() for first name, last name as the first name.